feat(exceptions): Integration global exception handling with structure API responses and log levels

mohamad-jaafar-monty opened the pull request:
# Description:
This PR introduces a comprehensive global exception handling system for the FastAPI backend. It standardizes how exceptions are raised, logged, and returned to clients across the application.

# Summary:
- `DevDoxAPIException` base class Centralizing error structure with:
 - `user_message`: shown to client
 - `log_message`: internal-only logs
 - `error_code`, `http_status`, `log_level`, `public_context`, `internal_context` 
- Custom exception subclass: `UnauthorizedAccess`
- Global exception handlers:
 - `devdox_base_exception_handler`
 - `http_exception_handler`
 - `generic_exception_handler`
- Dynamic log level support
 - `log_level` can be set to `warning`, `error`, or `exception`, enabling granular control over how errors are logged.
- Structured log message construction
 - Includes method, path, status, internal context (if present), and root exception trace (if applicable).
- Support for `debug` field
 - Automatically injected into APIResponse when in test/debug mode, excluded in production builds.

# Related Jira ticket:
- [DAPA-45](https://montyholding.atlassian.net/browse/DAPA-45)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a centralized, structured exception handling system with custom error classes for improved error responses and logging.
  - Added standardized error message constants for consistent user feedback across the application.
  - Enhanced API response serialization to better handle Pydantic models and nested data.

- **Bug Fixes**
  - Improved error message consistency and response formatting in endpoints and authentication flows.

- **Refactor**
  - Replaced legacy error constants and exception raising with new custom exception classes.
  - Centralized exception handler registration and updated imports for clarity and maintainability.
  - Reformatted code for improved readability and organization.

- **Tests**
  - Added comprehensive unit tests for the new exception handling and API response serialization.
  - Updated existing tests to align with the new error handling system and constants.

- **Style**
  - Improved code formatting and indentation for better readability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

[DAPA-45]: https://montyholding.atlassian.net/browse/DAPA-45?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
feat(exceptions): Integration global exception handling with structure API responses and log levels

# Description:
This PR introduces a comprehensive global exception handling system for the FastAPI backend. It standardizes how exceptions are raised, logged, and returned to clients across the application.

Summary:
- `DevDoxAPIException` base class Centralizing error structure with:
 - `user_message`: shown to client
 - `log_message`: internal-only logs
 - `error_code`, `http_status`, `log_level`, `public_context`, `internal_context`
- Custom exception subclass: `UnauthorizedAccess`
- Global exception handlers:
 - `devdox_base_exception_handler`
 - `http_exception_handler`
 - `generic_exception_handler`
- Dynamic log level support
 - `log_level` can be set to `warning`, `error`, or `exception`, enabling granular control over how errors are logged.
- Structured log message construction
 - Includes method, path, status, internal context (if present), and root exception trace (if applicable).
- Support for `debug` field
 - Automatically injected into APIResponse when in test/debug mode, excluded in production builds.

Commit:
test(unit): add complete coverage for exception handlers including all log level branches and subclass paths

# Summary

This PR introduces a comprehensive test suite that fully covers the application's global and custom exception handling logic under `app.exceptions.handlers`. Each handler path and condition has been thoroughly tested, including variations in log levels, context propagation, and subclass behavior.

# What was added

- A throwaway FastAPI app for isolated exception testing.
- Routes to simulate:
  - Generic `Exception`
  - `DevDoxAPIException` with `warning`, `error`, and `exception` log levels
  - `UnauthorizedAccess` subclass behavior
- Tests that cover:
  - Log output (`caplog`) verification
  - Status codes and response body structure
  - Inclusion/exclusion of `internal_context` and `public_context`
  - Log level–dependent output
- Parametrized and organized tests grouped by handler function, with docstrings for class and method clarity

Commit:
style: apply black code formatter to enforce consistent styling

mohamad-jaafar-monty commented on app/exceptions/register.py:
```diff
@@ -0,0 +1,28 @@
+
+def register_exception_handlers(app: FastAPI):
+	app.add_exception_handler(Exception, generic_exception_handler)
+	app.add_exception_handler(DevDoxAPIException, devdox_base_exception_handler)
```
Done

mohamad-jaafar-monty commented on tests/unit_test/app/utils/test_auth.py:
```diff
@@ -86,42 +87,42 @@ def test_successful_user_parsing(self):
-		assert INVALID_BEARER_TOKEN_SCHEMA in str(exc.value.detail)
+		assert exc.value.http_status == status.HTTP_401_UNAUTHORIZED
+		assert INVALID_BEARER_TOKEN_SCHEMA in exc.value.user_message
 	
```
Done

mohamad-jaafar-monty commented on tests/unit_test/app/utils/test_auth.py:
```diff
@@ -86,42 +87,42 @@ def test_successful_user_parsing(self):
-		assert INVALID_BEARER_TOKEN_SCHEMA in str(exc.value.detail)
+		assert exc.value.http_status == status.HTTP_401_UNAUTHORIZED
+		assert INVALID_BEARER_TOKEN_SCHEMA in exc.value.user_message
 	
```
Done

mohamad-jaafar-monty commented on tests/unit_test/app/utils/test_auth.py:
```diff
@@ -86,42 +87,42 @@ def test_successful_user_parsing(self):
-			assert exc.value.detail == constants.AUTH_FAILED
+			assert exc.value.http_status == status.HTTP_401_UNAUTHORIZED
+			assert app.exceptions.exception_constants.AUTH_FAILED in exc.value.user_message
 	
```
Done

mohamad-jaafar-monty commented on tests/unit_test/app/utils/test_auth.py:
```diff
@@ -86,42 +87,42 @@ def test_successful_user_parsing(self):
-			assert INVALID_BEARER_TOKEN_SCHEMA in str(exc.value.detail)
+			assert exc.value.http_status == status.HTTP_401_UNAUTHORIZED
+			assert INVALID_BEARER_TOKEN_SCHEMA in exc.value.user_message
 	
```
Done

mohamad-jaafar-monty commented on app/exceptions/handlers.py:
```diff
@@ -0,0 +1,85 @@
+    )
+
+
+def devdox_base_exception_handler(request: Request, exc: DevDoxAPIException):
```
Done

mohamad-jaafar-monty commented on app/exceptions/handlers.py:
```diff
@@ -0,0 +1,85 @@
+
+    logger.exception(
+        f"[UNHANDLED_EXCEPTION] {exc_type} occurred | Path: {path} | Method: {method} | Status: {status_code}"
+    )
```
Done

mohamad-jaafar-monty commented on tests/unit_test/app/exceptions/test_handlers.py:
```diff
@@ -0,0 +1,186 @@
+    def test_default_auth_message_included(self, fastapi_permissive_client):
+        """Checks that the default AUTH_FAILED message is used."""
+        resp = fastapi_permissive_client.get("/boom/unauth")
+        assert resp.json()["message"]
```
DONE

Commit:
coderabbit: added the coderabbit recommendations.

Hayat55 commented on app/exceptions/handlers.py:
```diff
@@ -0,0 +1,107 @@
+generic_exception_handler_status_code = status.HTTP_503_SERVICE_UNAVAILABLE
+
+
+def generic_exception_handler(request: Request, exc: Exception) -> dict[str, Any]:
```
as the return is  APIResponse.error so change the return type to JSONResponse

mohamad-jaafar-monty replied:
The type return by the handlers is dictated by the return type of `APIResponse`, which is set to return `dict[str, Any]`, to change the return type in the handlers, I have to make `APIResponse` return `JSONResponse`:
```python
from typing import Any, Dict, Optional

from fastapi.responses import JSONResponse

from app.config import settings


class APIResponse:
    ....

    @staticmethod
    def success(.....) -> Dict[str, Any]: # <===== Change this to -> JSONResponse
        .........
        return JSONResponse(content=response, status_code=200)

    @staticmethod
    def error(.......)-> Dict[str, Any]: # <===== Change this to -> JSONResponse
       .........
        return JSONResponse(content=response, status_code=status_code)

    @staticmethod
    def validation_error(....) -> Dict[str, Any]: # <===== Change this to -> JSONResponse
        .......
        return JSONResponse(content=response, status_code=422)


```

**What I will do:**
- Change APIResponse to return JSONResponse
- Change handlers to return JSONResponse

mohamad-jaafar-monty replied:
Working on it ....

Hayat55 commented on app/exceptions/handlers.py:
```diff
@@ -0,0 +1,107 @@
+    )
+
+
+def devdox_base_exception_handler(request: Request, exc: DevDoxAPIException):
```
specify the return type annotation

mohamad-jaafar-monty replied:
Working on it ....

Hayat55 commented on app/exceptions/handlers.py:
```diff
@@ -0,0 +1,107 @@
+    
+    path = request.url.path
+    method = getattr(request, "method", None) or request.scope.get("method") or "HTTP"
+    exc_type = type(exc).__name__
```
add the request id to help for tracing

Hayat55 commented on app/exceptions/handlers.py:
```diff
@@ -0,0 +1,107 @@
+    path = request.url.path
+    method = getattr(request, "method", None) or request.scope.get("method") or "HTTP"
+
+    log_parts = [
```
add the request if for tracing

Hayat55 commented on app/exceptions/handlers.py:
```diff
@@ -0,0 +1,107 @@
+
+    log_parts = [
+        f"[{exc.error_code}] {exc.log_message}",
+        f"Path: {path}",
```
add the user id if the user is logged in for tracing

Commit:
enhancement: Made error handling additionally consistent

# Changes Introduced:
- Handler return type now JSONResponse
- `debug` field is now handled in handlers as a central location instead of being scattered

Commit:
enhancement: Added trace_id middleware

# Changes Introduced:
- introduced the ability to generate a trace_id for every request
- trace_id will be in the logs and the header via `X-Trace-Id`

Commit:
enhancement: Added user_id to the logs and enhanced the logs

# Changes Introduced:
- introduced the ability to log user_id in the logs automatically, for authenticated users.

Commit:
fix: fixed tests after making get_current_user async

Commit:
style: applied black formatter for consistency

Commit:
fix: fixed some issues with rebase

Commit:
style: applied black formatter for consistency

Commit:
fix: fixed additional issues reported coderabbit

Commit:
feat: Added additional enhancements for Global Exception handler, a serializer, and an initial preparation for the implementation of the service/router layer separation

Summary:
- As part of the effort to start moving to have a service layer and a router layer where usually the service layer returns a python data structure. While the router layer returns a Json serialized value, I have introduced the ability to serialize pydantic BaseModels in api_response, and wrote tests for it

- Introduced additional Customized Exceptions

- Moved additional error constants from `app/utils/constants.py` to `app\exceptions\exception_constants.py` and fixed the locations in certain codes that require it.

Hayat55 commented on app/exceptions/custom_exceptions.py:
```diff
@@ -0,0 +1,131 @@
+        )
+
+
+class BadRequest(DevDoxAPIException):
```
Are we using this class ? same question for ResourceNotFound class

mohamad-jaafar-monty replied:
Yes but not in this pull request

Commit:
refactor: Removed custom logging of trace_id and user_id in preparation for integration with opentelemetry.

Commit:
style: Applied black formatter for consistency
